ydb/core/kqp/provider/rewrite_io_utils.cpp

```diff
@@ -0,0 +1,169 @@
+}
+
+void ChangeChild(const TExprNode::TPtr& parent, ui32 index, const TExprNode::TPtr& newChild) {
+    Y_ENSURE(parent->ChildrenSize() > index);
+
+    auto childrenToChange = parent->ChildrenList();
+    childrenToChange[index] = newChild;
+    parent->ChangeChildrenInplace(std::move(childrenToChange));
+}
+
+TExprNode::TPtr FindSavedQueryGraph(const TExprNode::TPtr& carrier) {
+    if (carrier->ChildrenSize() == 0) {
+        return nullptr;
+    }
+    auto lastChild = carrier->Children().back();
+    return lastChild->IsCallable(QueryGraphNodeSignature) ? lastChild->ChildPtr(0) : TExprNode::TPtr();
+}
+
+void SaveQueryGraph(const TExprNode::TPtr& carrier, TExprContext& ctx, const TExprNode::TPtr& payload) {
+    AddChild(carrier, ctx.NewCallable(payload->Pos(), QueryGraphNodeSignature, {payload}));
+}
+
+void InsertExecutionOrderDependencies(const TExprNode::TPtr& queryGraph, const TExprNode::TPtr& worldBefore) {
+    VisitExpr(
+        queryGraph,
+        nullptr,
+        [&worldBefore](const TExprNode::TPtr& node) {
+            if (node->ChildrenSize() > 0 && node->Child(0)->IsWorld()) {
+                ChangeChild(node, 0, worldBefore);
+            }
+            return true;
+        }
+    );
+}
+
+bool CheckTopLevelness(const TExprNode::TPtr& candidateRead, const TExprNode::TPtr& queryGraph) {
+    THashSet<TExprNode::TPtr> readsInCandidateSubgraph;
+    VisitExpr(candidateRead, [&readsInCandidateSubgraph](const TExprNode::TPtr& node) {
+        if (node->IsCallable(ReadName)) {
+            readsInCandidateSubgraph.emplace(node);
+        }
+        return true;
+    });
+
+    return !FindNode(queryGraph, [&readsInCandidateSubgraph](const TExprNode::TPtr& node) {
+        return node->IsCallable(ReadName) && !readsInCandidateSubgraph.contains(node);
+    });
+}
+
+TExprNode::TPtr FindTopLevelRead(const TExprNode::TPtr& queryGraph) {
+    const TExprNode::TPtr* lastReadInTopologicalOrder = nullptr;
+    VisitExpr(
+        queryGraph,
+        nullptr,
+        [&lastReadInTopologicalOrder](const TExprNode::TPtr& node) {
+            if (node->IsCallable(ReadName)) {
+                lastReadInTopologicalOrder = &node;
+            }
+            return true;
+        }
+    );
+
+    if (!lastReadInTopologicalOrder) {
+        return nullptr;
+    }
+
+    Y_ENSURE(CheckTopLevelness(*lastReadInTopologicalOrder, queryGraph),
+             "Info for developers: assumption that there is only one top level Read! is wrong\
+             for the expression graph of the query stored in the view:\n"
+                 << queryGraph->Dump());
+
+    return *lastReadInTopologicalOrder;
+}
+
+}
+
+TExprNode::TPtr RewriteReadFromView(
+    const TExprNode::TPtr& node,
+    TExprContext& ctx,
+    const TString& query,
+    const TString& cluster
+) {
+    const auto readNode = node->ChildPtr(0);
+    const auto worldBeforeThisRead = readNode->ChildPtr(0);
+
+    TExprNode::TPtr queryGraph = FindSavedQueryGraph(readNode);
```

Comment:
I save the query graph as the last child of the Read! from the view node to reuse it for both:
1) rewriting of the Left! callable from the Read! from the view node,
2) rewriting of the Right! callable from the Read! from the view node.

I need all the generated Read! (and other nodes) of the view graph to be the same ("same" = same object) for both Left! and Right! rewrites, because [CheckTx](https://github.com/ydb-platform/ydb/blob/c288b66eb979d88e47d5c141f1fe9f853336e08f/ydb/core/kqp/provider/yql_kikimr_opt_build.cpp#L658) expects all the Read! nodes to be seen on the path from the root node by the first child.